File: slang/Lexer.py
class TOKEN:
    ILLEGAL_TOKEN = -1
    TOK_PLUS = 0
    TOK_MUL = 1
    TOK_DIV = 2
    TOK_SUB = 3
    TOK_OPAREN = 4
    TOK_CPAREN = 5
    TOK_DOUBLE = 6
    TOK_NULL = 7


class Lexer:
    '''
    The Lexical Analysis Algorithm scans through the input
    and returns the token associated with the operator.
    '''
    def __init__(self, expr: str):
        self.iexpr = expr
        self.length = len(self.iexpr)
        self.index = 0

    def get_token(self):
        tok = TOKEN.TOK_NULL

        # skip white spaces
        while (self.index < self.length
               and (self.index == ' ' or self.iexpr[self.index] == '\t')):
            self.index += 1

        # end of string -> null
        if self.index == self.length:
            return TOKEN.TOK_NULL

        index_token = self.iexpr[self.index]

        if index_token == "+":
            tok = TOKEN.TOK_PLUS
            self.index += 1

        elif index_token == "-":
            tok = TOKEN.TOK_SUB
            self.index += 1

        elif index_token == "*":
            tok = TOKEN.TOK_MUL
            self.index += 1

        elif index_token == "/":
            tok = TOKEN.TOK_DIV
            self.index += 1

        elif index_token == "(":
            tok = TOKEN.TOK_OPAREN
            self.index += 1

        elif index_token == ")":
            tok = TOKEN.TOK_CPAREN
            self.index += 1

        # parsing string
        elif index_token.isdigit():
            num_string = ""

            while (self.index < self.length
                   and self.iexpr[self.index].isdigit()):

                num_string += self.iexpr[self.index]
                self.index += 1

            self.number = float(num_string)
            tok = TOKEN.TOK_DOUBLE

        else:
            raise ValueError("Error while Analysing the Token")

        return tok

    def get_number(self):
        return self.number

File: SlangJava/step-2/src/slang4java/Lexer/Lexer.java
package slang4java.Lexer;

public class Lexer {

    String iExpr;
    int index;
    double number;
    private int length;

    public Lexer(String iExpr) {
        this.iExpr = iExpr;
        length = iExpr.length();
        index = 0;
    }

    public TOKEN getToken() {
        TOKEN tok = TOKEN.ILLEGAL_TOKEN;

        while (index < length && iExpr.toCharArray()[index] == ' ' || index < length && iExpr.toCharArray()[index] == '\t')
            index++;

        // if end of string
        if (index == length) {
            return TOKEN.TOK_NULL;
        }

        switch (iExpr.toCharArray()[index]) {
            case '+' -> {
                tok = TOKEN.TOK_PLUS;
                index++;
                break;
            }
            case '-' -> {
                tok = TOKEN.TOK_SUB;
                index++;
                break;
            }
            case '*' -> {
                tok = TOKEN.TOK_MUL;
                index++;
                break;
            }
            case '/' -> {
                tok = TOKEN.TOK_DIV;
                index++;
                break;
            }
            case '(' -> {
                tok = TOKEN.TOK_OPREN;
                index++;
                break;
            }
            case ')' -> {
                tok = TOKEN.TOK_CPREN;
                index++;
                break;
            }
            default -> {
                if (Character.isDigit(iExpr.toCharArray()[index])) {
                    String str = "";
                    while (index < length && (iExpr.toCharArray()[index] == '0'
                            || iExpr.toCharArray()[index] == '1'
                            || iExpr.toCharArray()[index] == '2'
                            || iExpr.toCharArray()[index] == '3'
                            || iExpr.toCharArray()[index] == '4'
                            || iExpr.toCharArray()[index] == '5'
                            || iExpr.toCharArray()[index] == '6'
                            || iExpr.toCharArray()[index] == '7'
                            || iExpr.toCharArray()[index] == '8'
                            || iExpr.toCharArray()[index] == '9')) {
                        str += (iExpr.toCharArray()[index]);
                        index++;
                    }
                    number = Double.parseDouble(str);
                    tok = TOKEN.TOK_DOUBLE;
                } else {
                    throw new IllegalStateException("Unexpected value: " + iExpr.toCharArray()[index]);
                }
                //            default ->
            }
        }

        return tok;

    }

    public double getNumber() {
        return number;
    }
}
